File: src/types/fhir.ts
/**
 * TypeScript types mirroring the Rust FHIR structs.
 *
 * Field names use camelCase to match the Rust structs' #[serde(rename_all = "camelCase")].
 */

/** A FHIR R4 resource stored in the encrypted database. */
export interface FhirResource {
  id: string;
  resourceType: string;
  resource: Record<string, unknown>;
  versionId: number;
  lastUpdated: string;
  createdAt: string;
  updatedAt: string;
}

/** Input for creating a new FHIR resource. */
export interface CreateFhirResource {
  resourceType: string;
  resource: Record<string, unknown>;
}

/** Input for updating an existing FHIR resource. */
export interface UpdateFhirResource {
  id: string;
  resource: Record<string, unknown>;
}

/** Response wrapper for listing FHIR resources. */
export interface FhirResourceList {
  resources: FhirResource[];
  total: number;
}

/** Database health status from the check_db command. */
export interface DbStatus {
  encrypted: boolean;
  cipher_version: string;
  page_count: number;
}

/** Application info from the get_app_info command. */
export interface AppInfo {
  version: string;
  db_path: string;
}

File: src-tauri/src/db/models/fhir.rs
use serde::{Deserialize, Serialize};

/// A FHIR R4 resource stored in the encrypted database.
///
/// Fields use camelCase serialization to match Tauri 2's default
/// serde behavior for frontend interop.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FhirResource {
    pub id: String,
    pub resource_type: String,
    pub resource: serde_json::Value,
    pub version_id: i64,
    pub last_updated: String,
    pub created_at: String,
    pub updated_at: String,
}

/// Input for creating a new FHIR resource.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateFhirResource {
    pub resource_type: String,
    pub resource: serde_json::Value,
}

/// Input for updating an existing FHIR resource.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateFhirResource {
    pub id: String,
    pub resource: serde_json::Value,
}

/// Response wrapper for listing FHIR resources.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FhirResourceList {
    pub resources: Vec<FhirResource>,
    pub total: i64,
}
